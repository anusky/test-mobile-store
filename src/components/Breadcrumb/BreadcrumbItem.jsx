import ChevronRight from "../Icons/ChevronRight";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

const BreadcrumbItem = ({ crumb, isLast }) => (
  <>
    <li>
      {isLast ? (
        <span className="breadcrumb-item breadcrumb-item--last">
          {crumb.content}
        </span>
      ) : (
        <Link to={crumb?.url}>{crumb?.content}</Link>
      )}
    </li>
    {!isLast && (
      <li>
        <ChevronRight />
      </li>
    )}
  </>
);

export default BreadcrumbItem;
BreadcrumbItem.propTypes = {
  crumb: PropTypes.shape({
    url: PropTypes.string,
    content: PropTypes.node,
  }),
  isLast: PropTypes.bool.isRequired,
};
