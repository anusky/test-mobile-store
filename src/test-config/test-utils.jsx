import { render } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import PropTypes from "prop-types";

import productInfo from "./__fixtures__/product.json";
import productList from "./__fixtures__/productList.json";

const customRender = (ui, options = {}) => {
  const { routerInitialIndex = 0 } = options;

  const AllTheProviders = ({ children }) => {
    const routes = [
      {
        path: "/",
        element: children,
        loader: () => productList,
        handle: {
          crumb: () => ({ url: "/", content: "home route" }),
        },
        children: [
          {
            path: "/:productId",
            element: children,
            id: "product-page",
            loader: () => productInfo,
            handle: {
              crumb: () => ({ url: "/", content: "product route" }),
            },
          },
        ],
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ["/", "/123"],
      initialIndex: routerInitialIndex,
    });
    return <RouterProvider router={router}>{children}</RouterProvider>;
  };

  AllTheProviders.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return render(ui, { wrapper: AllTheProviders, ...options });
};

export * from "@testing-library/react";
export { customRender as render };
