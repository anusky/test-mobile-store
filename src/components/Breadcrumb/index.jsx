import { useLocation, useMatches } from "react-router-dom";
import BreadcrumbItem from "./BreadcrumbItem";

const Breadcrumb = () => {
  const location = useLocation();
  const matches = useMatches();

  const crumbs = matches
    .filter((match) => Boolean(match.handle?.crumb))
    .map((match) => match.handle.crumb(match.data));

  /**
   * We don't render any Breadcrumb when we are at home page
   */
  return location.pathname === "/" ? null : (
    <nav>
      <ol className="inline-flex gap-x-1 place-items-center">
        {crumbs.map((crumb, index) => (
          <BreadcrumbItem
            key={index}
            crumb={crumb}
            isLast={index === crumbs.length - 1}
          />
        ))}
      </ol>
    </nav>
  );
};

export default Breadcrumb;
