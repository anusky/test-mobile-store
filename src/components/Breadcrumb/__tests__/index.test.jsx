import { act } from "react-dom/test-utils";
import { render, screen } from "src/test-config/test-utils";
import Breadcrumb from "..";

describe("Breadcrumb coverage", () => {
  it("Should render navigation when the router points to /productId", async () => {
    render(<Breadcrumb />, { routerInitialIndex: 1 });

    expect(await screen.findByRole("navigation")).toBeInTheDocument();
  });

  it("Should render {home route > product route} navigation when the router points to /productId", async () => {
    render(<Breadcrumb />, { routerInitialIndex: 1 });

    const totalListItems = await screen.findAllByRole("listitem");
    expect(totalListItems.length).toBe(3);
    expect(await screen.findByText(/home route/i)).toBeInTheDocument();
    expect(await screen.findByText(/product route/i)).toBeInTheDocument();
  });

  it("Should not render any navigation on home route - default one", async () => {
    await act(() => {
      render(<Breadcrumb />);
    });
    expect(screen.queryByRole("navigation")).not.toBeInTheDocument();
  });

  it("matches snapshot with Home page route", async () => {
    let container;
    await act(() => {
      container = render(<Breadcrumb />);
    });
    expect(container.asFragment()).toMatchSnapshot();
  });
  it("matches snapshot with Product page route", async () => {
    let container;
    await act(() => {
      container = render(<Breadcrumb />, { routerInitialIndex: 1 });
    });
    expect(container.asFragment()).toMatchSnapshot();
  });
});
