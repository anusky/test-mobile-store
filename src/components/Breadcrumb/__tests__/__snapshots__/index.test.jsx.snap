// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb coverage matches snapshot with Home page route 1`] = `<DocumentFragment />`;

exports[`Breadcrumb coverage matches snapshot with Product page route 1`] = `
<DocumentFragment>
  <nav>
    <ol
      class="inline-flex gap-x-1 place-items-center"
    >
      <li>
        <a
          href="/"
        >
          home route
        </a>
      </li>
      <li>
        <svg
          class="w-5 h-5"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16.28 11.47a.75.75 0 010 1.06l-7.5 7.5a.75.75 0 01-1.06-1.06L14.69 12 7.72 5.03a.75.75 0 011.06-1.06l7.5 7.5z"
            fill-rule="evenodd"
          />
        </svg>
      </li>
      <li>
        <span
          class="breadcrumb-item breadcrumb-item--last"
        >
          product route
        </span>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;
